core/drivers/clk/clk-stm32mp25.c

```diff
@@ -0,0 +1,3446 @@
+
+	if (bypass || digbyp)
+		io_setbits32(address, BIT(bypass_data->bit_byp));
+}
+
+static void clk_oscillator_set_css(struct clk_stm32_priv *priv,
+				   struct clk_oscillator_data *osc_data,
+				   bool css)
+{
+	struct clk_stm32_css *css_data = osc_data->css;
+
+	if (css_data && css)
+		io_setbits32(priv->base + css_data->offset,
+			     BIT(css_data->bit_css));
+}
+
+static void clk_oscillator_set_drive(struct clk_stm32_priv *priv,
+				     struct clk_oscillator_data *osc_data,
+				     uint8_t lsedrv)
+{
+	struct clk_stm32_drive *drive_data = osc_data->drive;
+	uintptr_t address = 0;
+	uint32_t mask = 0;
+	uint32_t value = 0;
+
+	if (!drive_data)
+		return;
+
+	address = priv->base + drive_data->offset;
+
+	mask = SHIFT_U32(BIT(drive_data->drv_width) - 1, drive_data->drv_shift);
+
+	/*
+	 * Warning: not recommended to switch directly from "high drive"
+	 * to "medium low drive", and vice-versa.
+	 */
+	value = (io_read32(address) & mask) >> drive_data->drv_shift;
+
+	while (value != lsedrv) {
+		if (value > lsedrv)
+			value--;
+		else
+			value++;
+
+		io_clrsetbits32(address, mask,
+				SHIFT_U32(value, drive_data->drv_shift));
+	}
+}
+
+static void stm32_enable_oscillator_hse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_HSE);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_HSE];
+
+	if (!osci->freq)
+		return;
+
+	clk_oscillator_set_bypass(priv, osc_data, osci->digbyp, osci->bypass);
+
+	/* Enable clock and wait ready bit */
+	if (stm32_gate_rdy_enable(osc_data->gate_id))
+		panic("timeout to enable hse clock");
+
+	clk_oscillator_set_css(priv, osc_data, osci->css);
+}
+
+static void stm32_enable_oscillator_lse(struct clk_stm32_priv *priv,
+					struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_LSE);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSE];
+
+	if (!osci->freq)
+		return;
+
+	if (stm32_gate_is_enabled(osc_data->gate_id))
+		return;
+
+	clk_oscillator_set_bypass(priv, osc_data, osci->digbyp, osci->bypass);
+
+	clk_oscillator_set_drive(priv, osc_data, osci->drive);
+
+	/* Enable LSE clock, but don't wait ready bit */
+	stm32_gate_enable(osc_data->gate_id);
+}
+
+static void stm32_enable_oscillator_lsi(struct clk_stm32_priv *priv __unused,
+					struct stm32_clk_platdata *pdata)
+{
+	struct clk_oscillator_data *osc_data = clk_oscillator_get_data(OSC_LSI);
+	struct stm32_osci_dt_cfg *osci = &pdata->osci[OSC_LSI];
+
+	if (!osci->freq)
+		return;
+
+	/* Enable clock and wait ready bit */
+	if (stm32_gate_rdy_enable(osc_data->gate_id))
+		panic("timeout to enable lsi clock");
+}
+
+static TEE_Result clk_stm32_osc_msi_set_rate(struct clk_stm32_priv *priv,
+					     unsigned long rate)
+{
+	uintptr_t address = priv->base + RCC_BDCR;
+	uint32_t mask = RCC_BDCR_MSIFREQSEL;
+
+	switch (rate) {
+	case RCC_4_MHZ:
+		io_clrbits32(address, mask);
```

Comment:
I think here `io_setbits32_stm32shregs()`  and `io_clrbits32_stm32shregs()` would be safer, for concurrent access protection. It is not strictly needed now (all clock services use a common mutex protection) but safer if the common mutex is replaced with a per-clock protection mechanism.